CI: smoke test: insist on adding tests on PRs

edsantiago opened the pull request:
On each PR (other than CI:DOCS), check the list of git-touched
files, and abort if no tests are added. Include instructions
on how to bypass the check if tests really aren't needed.

Include a hardcoded exception list for PRs that only touch a
well-known subset of "safe" files: docs, .cirrus.yml, vendor.
This list is likely to need tuning, both right now and over time.

Add a test suite.

This is going to be controversial.

This is a github-only comment intended to bypass the new check: No Tests Needed

Signed-off-by: Ed Santiago <santiago@redhat.com>

edsantiago commented:
I did not include the magic 'no tests needed' string, so the first pass should fail smoke test. This is deliberate.

edsantiago commented:
Yay, [it failed](https://cirrus-ci.com/task/6031675102593024) (as expected):
```
 * f2a0009be "CI: smoke test: insist on adding tests on PRs" ... time="2021-01-07T21:20:39Z" level=info msg="[git] cmd: \"git --no-pager log --check f2a0009beb8148745ac8d2f23dc9bf46af43e56d^..f2a0009beb8148745ac8d2f23dc9bf46af43e56d -- . :(exclude)./vendor -- . :(exclude)docs/make.bat\""
PASS
pr-should-include-tests: PR does not include changes in the 'tests' directory

Please write a regression test for what you're fixing. Even if it
seems trivial or obvious, try to add a test that will prevent
regressions.

If your change is minor, feel free to piggyback on existing tests.
Every second counts in CI.

If your commit really, truly does not need tests, you can proceed
by adding 'No test needed' or 'No tests needed' to the body of
your commit message. Please think carefully before doing so.
```
I will now add the magic string to the github description and press Re-run.

edsantiago commented:
Interesting: updating the github message did not update `$CIRRUS_CHANGE_MESSAGE`. I must be misunderstanding how that works. Re-pushed with update to actual git commit, that worked.

@containers/podman-maintainers PTAL, and please give some thought to what other common cases we need to forgive.

Luap99 commented:
Not all tests are stored in the test directory. The unit test files are called `*_test.go` and are located next to the source code. You should accept these files as well.

edsantiago commented:
Good point, thank you. Fixed and pushed.

TomSweeneyRedHat commented on contrib/cirrus/pr-should-include-tests:
```diff
@@ -0,0 +1,64 @@
+Every second counts in CI.
+
+If your commit really, truly does not need tests, you can proceed
+by adding 'No test needed' or 'No tests needed' to the body of
```
Soft suggestion.  I'd change the required string to "[NO TESTS NEEDED]".  Hopefully, it will make the submitter think a bit more, and it will more likely stand out in the description for reviewers.

edsantiago replied:
I like that. The brackets are going to be hard to check for cleanly, so I will look into it next week. Thank you!

edsantiago replied:
Done and tested.

TomSweeneyRedHat commented:
I like the idea (may hate it later though when it bites me!  ;^) ), one suggestion for consideration.

vrothberg commented:
I'd love to have one "get out of jail" card similar to `[CI:DOCS]` that allows for merging without a test.  There are so many different kinds of changes that I fear we may miss one and have a red CI for an urgent change.  False positives can be annoying.

An alternative approach could be a GitHub action that drops a comment instead of blocking CI.

edsantiago commented:
You maybe missed the second sentence of the commit message?

vrothberg commented:
> You maybe missed the second sentence of the commit message?

I must have. Thanks :-)

LGTM

rhatdan commented:
Thanks @edsantiago I am going to hate this, but it is like taking your medicine, it is best for us all.
Should we wait to merge this until after 3.0 is released, though so we don't get hung up on this causing us a delay.

mheon commented on contrib/cirrus/pr-should-include-tests:
```diff
@@ -0,0 +1,64 @@
+                       fgrep -vx go.mod               |
+                       fgrep -vx go.sum               |
+                       egrep -v  '^docs/'             |
+                       egrep -v  '^vendor/')
```
Add to this: `version/`, `contrib/`, `hack/` - I need all of those for the release process, and I do not want to have to add tests for new releases.

mheon replied:
Further, anything that ends in `.md` in the base directory. We should not be requiring this for readme changes.

edsantiago replied:
Done; also changelog.txt.

rhatdan commented:
Also should ignore changes to version/version.go

Luap99 reviewed (commented):
Please document this in CONTRIBUTING.md and the PR template.

edsantiago commented:
>Please document this in CONTRIBUTING.md and the PR template

CONTRIBUTING.md : done, thanks for the suggestion. PR template: not done, I'm actually reluctant to do so. I fear it would give new contributors license to just add the magic string without even thinking. I prefer to have CI fail outright (it will do so quickly), this forces them to go through the effort of a new push; and, I hope, to actually consider adding tests.

edsantiago commented:
>Should we wait to merge this until after 3.0 is released

As soon as this merges, we're going to have a lot of angry devs. Release timing will increase that anger. So, yes, please defer until the team thinks most appropriate.

rhatdan commented:
I am fine with doing this after day 3.0 ships.
We will have to react quickly and be there to help contrinbutors who just don't have the time to write tests.

vrothberg commented:
Sounds good to me. Great work, @edsantiago!

rhatdan commented:
/hold
do not merge until post 3.0

edsantiago commented:
Rebased, and updated `CONTRIBUTING.md` with instructions on bypassing the block. I again looked at the [PR template](https://github.com/containers/podman/blob/8c6df5e93e6941a7f50da651678751f7dfec900e/.github/PULL_REQUEST_TEMPLATE.md) and again opt not to include instructions there, since it's really just a short-and-sweet link to `CONTRIBUTING.md`.

And, added `tests-included` Makefile target, now required by `validate`. I'm not 100% sure this one will work as desired, I think it may result in false positives, but there's really no way to test it other than in the field. Please let me know if you see this trigger inappropriately.

rhatdan commented:
/lgtm

rhatdan commented:
/hold cancel
